zio-http/src/main/scala/zio/http/netty/client/NettyConnectionPool.scala

```diff
@@ -189,25 +190,44 @@ object NettyConnectionPool {
       idleTimeout: Option[Duration],
       connectionTimeout: Option[Duration],
       localAddress: Option[InetSocketAddress] = None,
-    )(implicit trace: Trace): ZIO[Scope, Throwable, JChannel] =
+    )(implicit trace: Trace): ZIO[Scope, Throwable, JChannel] = {
+      val key = PoolKey(
+        location,
+        proxy,
+        sslOptions,
+        maxInitialLineLength,
+        maxHeaderSize,
+        decompression,
+        idleTimeout,
+        connectionTimeout,
+      )
+
       pool
-        .get(
-          PoolKey(
-            location,
-            proxy,
-            sslOptions,
-            maxInitialLineLength,
-            maxHeaderSize,
-            decompression,
-            idleTimeout,
-            connectionTimeout,
-          ),
-        )
+        .get(key)
+        .flatMap { channel =>
+          // Channel might have closed while in the pool, either because of a timeout or because of a connection error
+          // We retry to get another channel from the pool up to the number of minimum channels in the pool + 1
+          if (channel.isOpen) ZIO.succeed(channel)
+          else invalidate(channel) *> ZIO.fail(None)
+        }
+        .retry(retrySchedule(minSize(key) + 1))
```

Comment:
I'm not sure it makes sense to have this logic because we already know that the channel might become invalid immediately after we check it so we still need to fail if we use a channel that is invalid and so can handle all of the logic there.

Also if we do have something like this I don't think there is any logical relationship between the pool size and the number of times we should retry. This seems to be based on the idea that we're checking "all" the values in the pool before giving up but the items in the pool are potentially constantly changing so a fixed number of retries might make more sense here.

Reply:
Yeah I think you might be right. I was thinking of covering a potential edge case where all the channels in the pool are closed at the same time, and there is only a single thread executing a request. However, I think it's likely better for the behaviour of the pool to be more deterministic by having a fixed number of retries. I used 3 (for no specific reason) but I'm happy to change it to something lower / higher

Reply:
@adamgfraser I had to change the code back to having dynamic retries. I added a test that showcases that a fixed number of retries would cause issues in cases that all channels in the connection pool are closed (e.g., in case of server idle timeouts)

Reply:
> This seems to be based on the idea that we're checking "all" the values in the pool before giving up but the items in the pool are potentially constantly changing so a fixed number of retries might make more sense here.

PS: I agree with your point here, having the number of retries equal the max number of items that are possible to exist in the pool doesn't guarantee that we checked every single item. However, it does guarantee that the `get` effect of ZPool is executed at least once